docs/INTRO.md

```diff
@@ -0,0 +1,441 @@
+- `a :: PType` is short-hand for "Plutarch Type". We prefix these types with a capital `P`, such as `PInteger`, `PBool`, and so forth. _Tagging_ a `Term` with a `PType` indicates the type of the `Term`'s return value. Doing this allows us to bridge between the simple type system of Plutarch and the untyped UPLC -- the tags facilitate efficient conversions to and from the `Data` representation required by UPLC terms and the simple type system of Plutarch ` Term`'s. On the other hand, it also can provide an easier way to convert between Haskell values and Plutarch ` Term`'s via type synonyms.
+
+Note that we _should not_ think of a type of kind `PType` as carrying a value; it is a tag for a computation that may produce a value. For instance, the definition of `PInteger` is simply
+
+```hs
+data PInteger s
+```
+
+That is, there are no data constructors. _If_ a value of type `Term s PInteger` successfully executes the computation within context `s`, the value _computed_ will be an integer. We will never encounter values such as `y :: PInteger; y = 3`; they simply do not exist. While readers new to Plutarch may need some time to fit this into their mental model, it is a crucial distinction to keep in mind.
+
+> For brevity, we will say a "value of type `Term s a` will evaluate to (...)". This phrase will carry two implicit notions: one, that `Term s a` represents a computation executed in the context `s`; two, evaluating `Term s a` is not guaranteed to succeed.
+
+In brief, when writing Plutarch scripts, we have a few tasks:
+
+- A.) Defining _Plutarch Types_ (or _`PType`s_). We prefix these types with a capital `P`, such as `PInteger`, `PMaybe a`, `PBool`, and so forth. As previously mentioned, these form the "tags" for Plutarch `Term`'s, representing the type of the result of compiling and evaluating a Plutarch Script.
+- B.) Working with _Plutarch `Terms`_, which are values of the type `Term (s :: S) (a :: PType)`. These are the Plutarch scripts themselves, from which we build up more complex scripts before compiling and executing them on-chain.
+- C.) Writing Haskell-level functions _between Plutarch Terms_ (i.e., with types like `Term s a -> Term s b`). Doing so allows us to leverage Haskell's language features and ecosystem.
+- D.) Efficiently Converting the functions from _(C.)_ to _Plutarch-level functions_, which are of the type `Term s (a :--> b)`. We can _directly_ convert the functions from (C.) to Plutarch-level functions at the most naive level using `plam`. Additional Plutarch utilities provide for optimization opportunities.
+- E.) Compiling and executing the functions from _(D.)_, targetting UPLC for on-chain usage. 
+
+As a preview, the bridge Plutarch provides between Haskell and UPLC looks something like this:
+
+```
+------------------------------------------------------
```

Comment:
Yeah, I wasn't super happy with this chart myself. I'll revisit these suggestions later and try to draw up something better. Happy to incorporate more suggestions if you think of them

Reply:
Removing the `pfromData/pdata` part is enough in my eyes, as those are nothing special and don't establish any bridges between worlds. You can fill that gap with `compile` if you want - that fills the role of "Plutarch -> UPLC". I'm not too sure about the reverse - @blamario is working on FFI that might achieve such a relation eventually.

You *could* perhaps have another box showcasing Haskell functions working on `Term`s - these can be converted to Plutarch term world using `plam`, of course. The reverse relation is also possible, something like:
```hs
pf :: Term s (a :--> b :--> c)

-- recovering a haskell level function from a plutarch function
hf :: Term s a -> Term s b -> Term s c
hf x y = pf # x # y
```

I don't think that box is necessary though - I personally don't like treating functions and regular values differently. But if you believe it could potentially help users - I have absolutely nothing against it.

Reply:
Edited the diagram and added your example. I think its as good a place as any to introduce the convention the `h` indicates a function-between-terms and `p` indicates a lambda term